wakebox/main.cpp

```diff
@@ -32,57 +33,120 @@
+// Wakebox does not provide direct control of the command running dir at this time.
+const std::string command_running_dir(
+	const std::vector<mount_op>& mount_ops,
+	const std::string& host_workspace_dir,
+	const std::string& subdir)
 {
-	struct option options[] {
-		{'p', "params",       GOPT_ARGUMENT_REQUIRED},
-		{'o', "output-stats", GOPT_ARGUMENT_REQUIRED},
-		{'i', "interactive",  GOPT_ARGUMENT_FORBIDDEN},
-		{'s', "force-shell",  GOPT_ARGUMENT_FORBIDDEN},
-		{'h', "help",         GOPT_ARGUMENT_FORBIDDEN},
-		{0,   0,              GOPT_LAST }
-	};
+	// If we have a workspace mount, we want to default to that location.
+	for (auto &x : mount_ops) {
+		if (x.type == "workspace") {
+			if (x.destination.at(0) == '/')
+				return x.destination + "/" + subdir;
+			else
+				// convert a workspace relative path into absolute path
+				return host_workspace_dir + "/" + x.destination + "/" + subdir;
+		}
+	}
+	// If we're binding in the parent namespace's current working directory, use that.
+	for (auto &x : mount_ops)
+		if (x.type == "bind" && host_workspace_dir == x.source)
+			return x.destination + "/" + subdir;
 
-	argc = gopt(argv, options);
-	gopt_errors("wakebox", options);
+	// If we have a replacement rootfs, we know we atleast have "/".
+	for (auto &x : mount_ops)
+		if (x.destination == "/")
+			return "/" + subdir;
 
-	bool has_output = arg(options, "output-stats")->count > 0;
-	bool use_stdin_file = arg(options, "interactive")->count == 0;
-	bool use_shell = arg(options, "force-shell")->count > 0;
+	// Try the current directory, which should exist if we have no replacement rootfs.
+	return host_workspace_dir + "/" + subdir;
+}
 
-	const char* params  = arg(options, "params" )->argument;
-	const char* result_path = arg(options, "output-stats")->argument;
+int run_interactive(
+	const std::string &rootfs,
+	const std::vector<std::string> toolchains,
+	const std::vector<mount_op> binds,
+	const std::vector<std::string> command
+){
+	struct fuse_args fa(get_cwd(), false);
+	fa.command = command;
 
-	if (arg(options, "help")->count > 0 || !params) {
-		print_help();
-		return 1;
+	// Interactive mode does not provide userid control at this time.
+	fa.userid = 0;
+	fa.groupid = 0;
+
+	if (!rootfs.empty())
+		fa.mount_ops.push_back({"squashfs", rootfs, "/", false});
+
+	for (auto &tool : toolchains)
+		fa.mount_ops.push_back({"squashfs", tool, "", false});
+
+	fa.mount_ops.insert(fa.mount_ops.end(), binds.begin(), binds.end());
+
+	if (rootfs.empty()) {
+		fa.environment.push_back(std::string("HOME=") + std::getenv("HOME"));
+		fa.environment.push_back(std::string("USER=") + std::getenv("USER"));
 	}
 
+	const char *term = std::getenv("TERM");
+	fa.environment.push_back(std::string("TERM=") + term);
+
+	fa.command_running_dir = command_running_dir(fa.mount_ops, fa.working_dir, "");
+
+	std::string result;
+	return run_in_fuse(fa, result);
+}
+
+
+int run_batch(
+	const char* params_path,
```

Comment:
You could have used `const std::string &` and it would auto-convert the type at call sites.